File: docs/examples/python/quickstart.py
#!/usr/bin/env python3
"""
QES Platform Python SDK Quick Start Example

This example demonstrates how to:
1. Authenticate with the QES Platform API
2. Authenticate a user with a QES provider (Freja eID)
3. Sign a PDF document with qualified electronic signature
4. Verify the created signature

Prerequisites:
- QES Platform account and API credentials
- Freja eID test account (or other supported provider)
- Python 3.8+ with requests library

Install dependencies:
    pip install requests cryptography

Usage:
    python quickstart.py
"""

import os
import json
import time
import base64
from typing import Dict, Any, Optional
import requests
from pathlib import Path


class QESPlatformClient:
    """Simple QES Platform API client for demonstration purposes."""
    
    def __init__(self, api_url: str, api_key: str, tenant_id: str):
        self.api_url = api_url.rstrip('/')
        self.api_key = api_key
        self.tenant_id = tenant_id
        self.session = requests.Session()
        
        # Set default headers
        self.session.headers.update({
            'Authorization': f'Bearer {api_key}',
            'X-Tenant-ID': tenant_id,
            'Content-Type': 'application/json',
            'User-Agent': 'QES-Platform-Python-Example/1.0'
        })
    
    def health_check(self) -> Dict[str, Any]:
        """Check API connectivity and health."""
        response = self.session.get(f'{self.api_url}/health')
        response.raise_for_status()
        return response.json()
    
    def list_providers(self) -> Dict[str, Any]:
        """List available QES providers."""
        response = self.session.get(f'{self.api_url}/providers')
        response.raise_for_status()
        return response.json()
    
    def authenticate_user(self, provider: str, user_identifier: str, 
                         redirect_uri: str, **kwargs) -> Dict[str, Any]:
        """Initiate user authentication with QES provider."""
        payload = {
            'provider': provider,
            'user_identifier': user_identifier,
            'auth_method': 'oauth2',
            'redirect_uri': redirect_uri,
            **kwargs
        }
        
        response = self.session.post(f'{self.api_url}/auth/login', json=payload)
        response.raise_for_status()
        return response.json()
    
    def handle_auth_callback(self, provider: str, session_id: str, 
                           callback_params: Dict[str, Any]) -> Dict[str, Any]:
        """Handle authentication callback from QES provider."""
        payload = {
            'provider': provider,
            'session_id': session_id,
            'callback_params': callback_params
        }
        
        response = self.session.post(f'{self.api_url}/auth/callback', json=payload)
        response.raise_for_status()
        return response.json()
    
    def list_certificates(self) -> Dict[str, Any]:
        """List available certificates for signing."""
        response = self.session.get(f'{self.api_url}/certificates')
        response.raise_for_status()
        return response.json()
    
    def sign_document(self, document_path: str, document_name: str,
                     signature_format: str, certificate_id: Optional[str] = None,
                     **kwargs) -> Dict[str, Any]:
        """Sign a document with qualified electronic signature."""
        
        # Prepare multipart form data
        files = {
            'document': (document_name, open(document_path, 'rb'), 'application/pdf')
        }
        
        data = {
            'document_name': document_name,
            'signature_format': signature_format,
            **kwargs
        }
        
        if certificate_id:
            data['certificate_id'] = certificate_id
        
        # Remove Content-Type header for multipart requests
        headers = self.session.headers.copy()
        headers.pop('Content-Type', None)
        
        response = self.session.post(
            f'{self.api_url}/sign',
            files=files,
            data=data,
            headers=headers
        )
        response.raise_for_status()
        return response.json()
    
    def verify_signature(self, signed_document_path: str, 
                        verification_level: str = 'qualified') -> Dict[str, Any]:
        """Verify a digitally signed document."""
        
        files = {
            'signed_document': (
                'signed_document.pdf',
                open(signed_document_path, 'rb'),
                'application/pdf'
            )
        }
        
        data = {
            'verification_level': verification_level
        }
        
        # Remove Content-Type header for multipart requests
        headers = self.session.headers.copy()
        headers.pop('Content-Type', None)
        
        response = self.session.post(
            f'{self.api_url}/verify',
            files=files,
            data=data,
            headers=headers
        )
        response.raise_for_status()
        return response.json()
    
    def get_usage_stats(self, period: str = 'day') -> Dict[str, Any]:
        """Get tenant usage statistics."""
        response = self.session.get(f'{self.api_url}/tenant/usage?period={period}')
        response.raise_for_status()
        return response.json()


def create_sample_pdf() -> str:
    """Create a simple PDF document for testing."""
    try:
        from reportlab.pdfgen import canvas
        from reportlab.lib.pagesizes import letter
        
        filename = 'sample_document.pdf'
        c = canvas.Canvas(filename, pagesize=letter)
        
        # Add content to PDF
        c.drawString(100, 750, "QES Platform Test Document")
        c.drawString(100, 700, "This is a sample document for digital signature testing.")
        c.drawString(100, 650, f"Generated on: {time.strftime('%Y-%m-%d %H:%M:%S')}")
        c.drawString(100, 600, "Content: Lorem ipsum dolor sit amet, consectetur adipiscing elit.")
        
        c.save()
        return filename
        
    except ImportError:
        # Fallback: create a minimal PDF manually
        print("ReportLab not available, creating minimal PDF...")
        
        minimal_pdf = b"""%PDF-1.4
1 0 obj
<<
/Type /Catalog
/Pages 2 0 R
>>
endobj

2 0 obj
<<
/Type /Pages
/Kids [3 0 R]
/Count 1
>>
endobj

3 0 obj
<<
/Type /Page
/Parent 2 0 R
/MediaBox [0 0 612 792]
/Contents 4 0 R
>>
endobj

4 0 obj
<<
/Length 55
>>
stream
BT
/F1 12 Tf
100 700 Td
(QES Platform Test Document) Tj
ET
endstream
endobj

xref
0 5
0000000000 65535 f 
0000000009 00000 n 
0000000058 00000 n 
0000000115 00000 n 
0000000206 00000 n 
trailer
<<
/Size 5
/Root 1 0 R
>>
startxref
312
%%EOF"""
        
        filename = 'sample_document.pdf'
        with open(filename, 'wb') as f:
            f.write(minimal_pdf)
        return filename


def demonstrate_workflow():
    """Demonstrate complete QES Platform workflow."""
    
    # Load configuration from environment variables
    api_url = os.getenv('QES_API_URL', 'http://localhost:8000/api/v1')
    api_key = os.getenv('QES_API_KEY', 'dev-api-key')
    tenant_id = os.getenv('QES_TENANT_ID', 'dev-tenant')
    
    if not all([api_url, api_key, tenant_id]):
        print("❌ Missing required environment variables:")
        print("   QES_API_URL, QES_API_KEY, QES_TENANT_ID")
        print("\nSet them like this:")
        print("   export QES_API_URL='https://your-tenant.qes-platform.com/api/v1'")
        print("   export QES_API_KEY='your-api-key'")
        print("   export QES_TENANT_ID='your-tenant-id'")
        return
    
    print("🚀 QES Platform Python SDK Quick Start")
    print("=" * 50)
    
    # Initialize client
    client = QESPlatformClient(api_url, api_key, tenant_id)
    
    try:
        # 1. Health Check
        print("\n1️⃣  Checking API connectivity...")
        health = client.health_check()
        print(f"   ✅ API Status: {health.get('status', 'unknown')}")
        print(f"   📅 Timestamp: {health.get('timestamp', 'unknown')}")
        
        # 2. List Available Providers
        print("\n2️⃣  Listing available QES providers...")
        providers = client.list_providers()
        print(f"   📋 Found {len(providers.get('providers', []))} providers:")
        for provider in providers.get('providers', []):
            status = "🟢" if provider.get('is_available') else "🔴"
            print(f"      {status} {provider.get('name')} ({provider.get('country_code')})")
        
        # 3. Create Sample Document
        print("\n3️⃣  Creating sample document...")
        sample_pdf = create_sample_pdf()
        print(f"   📄 Created: {sample_pdf}")
        
        # 4. User Authentication (Demo Mode)
        print("\n4️⃣  Initiating user authentication...")
        
        # In demo mode, we'll simulate the authentication flow
        if 'localhost' in api_url or 'dev' in tenant_id:
            print("   🧪 Demo Mode: Simulating authentication...")
            auth_result = {
                'auth_url': 'https://demo.frejaeid.com/auth?state=demo123',
                'session_id': 'demo_session_123',
                'state': 'demo123'
            }
            print(f"   🔗 Auth URL: {auth_result['auth_url']}")
            
            # Simulate callback
            print("   ⏳ Simulating user authentication...")
            time.sleep(2)
            
            callback_result = {
                'status': 'authenticated',
                'access_token': 'demo_access_token_123',
                'user_info': {
                    'user_id': 'demo_user_123',
                    'given_name': 'John',
                    'family_name': 'Doe',
                    'email': 'john.doe@example.com',
                    'country_code': 'SE'
                }
            }
            print(f"   ✅ User authenticated: {callback_result['user_info']['given_name']} {callback_result['user_info']['family_name']}")
            
        else:
            # Real authentication flow
            auth_result = client.authenticate_user(
                provider='freja-se',
                user_identifier='user@example.com',
                redirect_uri='https://your-app.com/callback'
            )
            print(f"   🔗 Please visit: {auth_result.get('auth_url')}")
            print("   ⏳ Waiting for authentication...")
            
            # In a real application, you would handle the callback
            # For this demo, we'll wait for user input
            input("   👆 Press Enter after completing authentication...")
        
        # 5. List Available Certificates
        print("\n5️⃣  Listing available certificates...")
        try:
            certificates = client.list_certificates()
            cert_list = certificates.get('certificates', [])
            
            if cert_list:
                print(f"   📜 Found {len(cert_list)} certificates:")
                for cert in cert_list[:3]:  # Show first 3
                    print(f"      🔐 {cert.get('subject_dn', 'Unknown')}")
                    print(f"         Valid: {cert.get('valid_from')} → {cert.get('valid_to')}")
                
                certificate_id = cert_list[0].get('certificate_id')
            else:
                print("   ⚠️  No certificates found, using demo certificate")
                certificate_id = 'demo_cert_123'
                
        except Exception as e:
            print(f"   ⚠️  Certificate listing failed: {e}")
            certificate_id = 'demo_cert_123'
        
        # 6. Sign Document
        print("\n6️⃣  Signing document...")
        signature_result = client.sign_document(
            document_path=sample_pdf,
            document_name='sample_contract.pdf',
            signature_format='PAdES-LTA',
            certificate_id=certificate_id
        )
        
        print(f"   ✅ Signature created successfully!")
        print(f"   🆔 Signature ID: {signature_result.get('signature_id')}")
        print(f"   📝 Format: {signature_result.get('signature_format')}")
        print(f"   ⏰ Timestamp: {signature_result.get('timestamp')}")
        
        # Save signed document
        if 'signed_document' in signature_result:
            signed_doc_data = base64.b64decode(signature_result['signed_document'])
            signed_filename = 'signed_document.pdf'
            with open(signed_filename, 'wb') as f:
                f.write(signed_doc_data)
            print(f"   💾 Saved signed document: {signed_filename}")
        
        # 7. Verify Signature
        if 'signed_document' in signature_result:
            print("\n7️⃣  Verifying signature...")
            verification_result = client.verify_signature(signed_filename)
            
            is_valid = verification_result.get('is_valid', False)
            status_icon = "✅" if is_valid else "❌"
            print(f"   {status_icon} Signature Valid: {is_valid}")
            print(f"   🔍 Verification Level: {verification_result.get('verification_level')}")
            
            validation_details = verification_result.get('validation_details', {})
            print(f"   📋 Certificate Valid: {validation_details.get('certificate_valid')}")
            print(f"   🔒 Signature Intact: {validation_details.get('signature_intact')}")
            print(f"   ⏱️  Timestamp Valid: {validation_details.get('timestamp_valid')}")
        
        # 8. Usage Statistics
        print("\n8️⃣  Checking usage statistics...")
        try:
            usage_stats = client.get_usage_stats()
            print(f"   📊 Signatures Created Today: {usage_stats.get('signatures_created', 0)}")
            print(f"   ✔️  Signatures Verified Today: {usage_stats.get('signatures_verified', 0)}")
            print(f"   📈 API Requests Today: {usage_stats.get('api_requests', 0)}")
        except Exception as e:
            print(f"   ⚠️  Usage stats unavailable: {e}")
        
        print("\n🎉 QES Platform workflow completed successfully!")
        print("\n📚 Next Steps:")
        print("   • Explore the full API documentation")
        print("   • Integrate with your application")
        print("   • Test with different signature formats")
        print("   • Set up webhook notifications")
        
    except requests.exceptions.RequestException as e:
        print(f"\n❌ API Error: {e}")
        if hasattr(e, 'response') and e.response is not None:
            try:
                error_data = e.response.json()
                print(f"   Error Details: {error_data.get('message', 'Unknown error')}")
            except:
                print(f"   HTTP Status: {e.response.status_code}")
        
    except Exception as e:
        print(f"\n❌ Unexpected Error: {e}")
        
    finally:
        # Cleanup
        for file in ['sample_document.pdf', 'signed_document.pdf']:
            if os.path.exists(file):
                os.remove(file)
                print(f"   🧹 Cleaned up: {file}")


if __name__ == '__main__':
    print("QES Platform Python SDK - Quick Start Example")
    print("=" * 50)
    print()
    print("This example demonstrates the complete QES workflow:")
    print("• API connectivity check")
    print("• Provider listing")
    print("• User authentication")
    print("• Document signing")
    print("• Signature verification")
    print()
    
    # Check for environment setup
    missing_vars = []
    for var in ['QES_API_URL', 'QES_API_KEY', 'QES_TENANT_ID']:
        if not os.getenv(var):
            missing_vars.append(var)
    
    if missing_vars:
        print("⚠️  Missing environment variables for production use:")
        for var in missing_vars:
            print(f"   • {var}")
        print("\nRunning in demo mode with local development server...")
        print("For production, set the above environment variables.")
        print()
    
    input("Press Enter to start the demonstration...")
    demonstrate_workflow()

File: docs/examples/javascript/quickstart.js
#!/usr/bin/env node

/**
 * QES Platform JavaScript SDK Quick Start Example
 * 
 * This example demonstrates how to:
 * 1. Authenticate with the QES Platform API
 * 2. Authenticate a user with a QES provider
 * 3. Sign a document with qualified electronic signature
 * 4. Verify the created signature
 * 
 * Prerequisites:
 * - QES Platform account and API credentials
 * - Node.js 16+ with npm
 * - Supported QES provider account
 * 
 * Install dependencies:
 *   npm install axios form-data fs-extra
 * 
 * Usage:
 *   node quickstart.js
 */

const axios = require('axios');
const FormData = require('form-data');
const fs = require('fs-extra');
const path = require('path');

/**
 * Simple QES Platform API client for demonstration purposes
 */
class QESPlatformClient {
  constructor(apiUrl, apiKey, tenantId) {
    this.apiUrl = apiUrl.replace(/\/$/, '');
    this.apiKey = apiKey;
    this.tenantId = tenantId;
    
    // Configure axios defaults
    this.client = axios.create({
      baseURL: this.apiUrl,
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'X-Tenant-ID': tenantId,
        'Content-Type': 'application/json',
        'User-Agent': 'QES-Platform-JS-Example/1.0'
      },
      timeout: 30000 // 30 second timeout
    });
    
    // Add response interceptor for error handling
    this.client.interceptors.response.use(
      response => response,
      error => {
        if (error.response) {
          // Server responded with error status
          const errorMessage = error.response.data?.message || error.message;
          throw new Error(`API Error (${error.response.status}): ${errorMessage}`);
        } else if (error.request) {
          // Request made but no response
          throw new Error('Network Error: No response from server');
        } else {
          // Something else happened
          throw new Error(`Request Error: ${error.message}`);
        }
      }
    );
  }

  /**
   * Check API connectivity and health
   */
  async healthCheck() {
    const response = await this.client.get('/health');
    return response.data;
  }

  /**
   * List available QES providers
   */
  async listProviders() {
    const response = await this.client.get('/providers');
    return response.data;
  }

  /**
   * Initiate user authentication with QES provider
   */
  async authenticateUser(provider, userIdentifier, redirectUri, options = {}) {
    const payload = {
      provider,
      user_identifier: userIdentifier,
      auth_method: 'oauth2',
      redirect_uri: redirectUri,
      ...options
    };
    
    const response = await this.client.post('/auth/login', payload);
    return response.data;
  }

  /**
   * Handle authentication callback from QES provider
   */
  async handleAuthCallback(provider, sessionId, callbackParams) {
    const payload = {
      provider,
      session_id: sessionId,
      callback_params: callbackParams
    };
    
    const response = await this.client.post('/auth/callback', payload);
    return response.data;
  }

  /**
   * List available certificates for signing
   */
  async listCertificates() {
    const response = await this.client.get('/certificates');
    return response.data;
  }

  /**
   * Sign a document with qualified electronic signature
   */
  async signDocument(documentPath, documentName, signatureFormat, certificateId = null, options = {}) {
    const formData = new FormData();
    
    // Add document file
    const documentStream = fs.createReadStream(documentPath);
    formData.append('document', documentStream, {
      filename: documentName,
      contentType: 'application/pdf'
    });
    
    // Add form fields
    formData.append('document_name', documentName);
    formData.append('signature_format', signatureFormat);
    
    if (certificateId) {
      formData.append('certificate_id', certificateId);
    }
    
    // Add additional options
    Object.entries(options).forEach(([key, value]) => {
      formData.append(key, value);
    });
    
    const response = await this.client.post('/sign', formData, {
      headers: {
        ...formData.getHeaders(),
        // Remove default Content-Type to let axios set it with boundary
        'Content-Type': undefined
      }
    });
    
    return response.data;
  }

  /**
   * Verify a digitally signed document
   */
  async verifySignature(signedDocumentPath, verificationLevel = 'qualified') {
    const formData = new FormData();
    
    // Add signed document file
    const documentStream = fs.createReadStream(signedDocumentPath);
    formData.append('signed_document', documentStream, {
      filename: 'signed_document.pdf',
      contentType: 'application/pdf'
    });
    
    formData.append('verification_level', verificationLevel);
    
    const response = await this.client.post('/verify', formData, {
      headers: {
        ...formData.getHeaders(),
        'Content-Type': undefined
      }
    });
    
    return response.data;
  }

  /**
   * Get tenant usage statistics
   */
  async getUsageStats(period = 'day') {
    const response = await this.client.get(`/tenant/usage?period=${period}`);
    return response.data;
  }
}

/**
 * Create a simple PDF document for testing
 */
async function createSamplePdf() {
  const pdfContent = `%PDF-1.4
1 0 obj
<<
/Type /Catalog
/Pages 2 0 R
>>
endobj

2 0 obj
<<
/Type /Pages
/Kids [3 0 R]
/Count 1
>>
endobj

3 0 obj
<<
/Type /Page
/Parent 2 0 R
/MediaBox [0 0 612 792]
/Contents 4 0 R
/Resources <<
/Font <<
/F1 <<
/Type /Font
/Subtype /Type1
/BaseFont /Helvetica
>>
>>
>>
>>
endobj

4 0 obj
<<
/Length 120
>>
stream
BT
/F1 12 Tf
100 700 Td
(QES Platform Test Document) Tj
0 -20 Td
(Generated: ${new Date().toISOString()}) Tj
0 -20 Td
(This is a sample document for digital signature testing.) Tj
ET
endstream
endobj

xref
0 5
0000000000 65535 f 
0000000009 00000 n 
0000000058 00000 n 
0000000115 00000 n 
0000000300 00000 n 
trailer
<<
/Size 5
/Root 1 0 R
>>
startxref
470
%%EOF`;

  const filename = 'sample_document.pdf';
  await fs.writeFile(filename, pdfContent);
  return filename;
}

/**
 * Demonstrate complete QES Platform workflow
 */
async function demonstrateWorkflow() {
  // Load configuration from environment variables
  const apiUrl = process.env.QES_API_URL || 'http://localhost:8000/api/v1';
  const apiKey = process.env.QES_API_KEY || 'dev-api-key';
  const tenantId = process.env.QES_TENANT_ID || 'dev-tenant';
  
  if (!apiUrl || !apiKey || !tenantId) {
    console.log('❌ Missing required environment variables:');
    console.log('   QES_API_URL, QES_API_KEY, QES_TENANT_ID');
    console.log('\nSet them like this:');
    console.log('   export QES_API_URL="https://your-tenant.qes-platform.com/api/v1"');
    console.log('   export QES_API_KEY="your-api-key"');
    console.log('   export QES_TENANT_ID="your-tenant-id"');
    return;
  }

  console.log('🚀 QES Platform JavaScript SDK Quick Start');
  console.log('='.repeat(50));

  // Initialize client
  const client = new QESPlatformClient(apiUrl, apiKey, tenantId);

  try {
    // 1. Health Check
    console.log('\n1️⃣  Checking API connectivity...');
    const health = await client.healthCheck();
    console.log(`   ✅ API Status: ${health.status || 'unknown'}`);
    console.log(`   📅 Timestamp: ${health.timestamp || 'unknown'}`);

    // 2. List Available Providers
    console.log('\n2️⃣  Listing available QES providers...');
    const providers = await client.listProviders();
    const providerList = providers.providers || [];
    console.log(`   📋 Found ${providerList.length} providers:`);
    
    providerList.forEach(provider => {
      const status = provider.is_available ? '🟢' : '🔴';
      console.log(`      ${status} ${provider.name} (${provider.country_code})`);
    });

    // 3. Create Sample Document
    console.log('\n3️⃣  Creating sample document...');
    const samplePdf = await createSamplePdf();
    console.log(`   📄 Created: ${samplePdf}`);

    // 4. User Authentication (Demo Mode)
    console.log('\n4️⃣  Initiating user authentication...');
    
    let authResult, callbackResult;
    
    if (apiUrl.includes('localhost') || tenantId.includes('dev')) {
      console.log('   🧪 Demo Mode: Simulating authentication...');
      authResult = {
        auth_url: 'https://demo.frejaeid.com/auth?state=demo123',
        session_id: 'demo_session_123',
        state: 'demo123'
      };
      console.log(`   🔗 Auth URL: ${authResult.auth_url}`);
      
      // Simulate callback
      console.log('   ⏳ Simulating user authentication...');
      await new Promise(resolve => setTimeout(resolve, 2000));
      
      callbackResult = {
        status: 'authenticated',
        access_token: 'demo_access_token_123',
        user_info: {
          user_id: 'demo_user_123',
          given_name: 'John',
          family_name: 'Doe',
          email: 'john.doe@example.com',
          country_code: 'SE'
        }
      };
      console.log(`   ✅ User authenticated: ${callbackResult.user_info.given_name} ${callbackResult.user_info.family_name}`);
      
    } else {
      // Real authentication flow
      authResult = await client.authenticateUser(
        'freja-se',
        'user@example.com',
        'https://your-app.com/callback'
      );
      console.log(`   🔗 Please visit: ${authResult.auth_url}`);
      console.log('   ⏳ Waiting for authentication...');
      
      // In a real application, you would handle the callback
      console.log('   👆 Complete authentication in your browser and press Enter...');
      await new Promise(resolve => {
        process.stdin.once('data', resolve);
      });
    }

    // 5. List Available Certificates
    console.log('\n5️⃣  Listing available certificates...');
    let certificateId;
    
    try {
      const certificates = await client.listCertificates();
      const certList = certificates.certificates || [];
      
      if (certList.length > 0) {
        console.log(`   📜 Found ${certList.length} certificates:`);
        certList.slice(0, 3).forEach(cert => {
          console.log(`      🔐 ${cert.subject_dn || 'Unknown'}`);
          console.log(`         Valid: ${cert.valid_from} → ${cert.valid_to}`);
        });
        
        certificateId = certList[0].certificate_id;
      } else {
        console.log('   ⚠️  No certificates found, using demo certificate');
        certificateId = 'demo_cert_123';
      }
      
    } catch (error) {
      console.log(`   ⚠️  Certificate listing failed: ${error.message}`);
      certificateId = 'demo_cert_123';
    }

    // 6. Sign Document
    console.log('\n6️⃣  Signing document...');
    const signatureResult = await client.signDocument(
      samplePdf,
      'sample_contract.pdf',
      'PAdES-LTA',
      certificateId
    );
    
    console.log('   ✅ Signature created successfully!');
    console.log(`   🆔 Signature ID: ${signatureResult.signature_id}`);
    console.log(`   📝 Format: ${signatureResult.signature_format}`);
    console.log(`   ⏰ Timestamp: ${signatureResult.timestamp}`);
    
    // Save signed document
    let signedFilename;
    if (signatureResult.signed_document) {
      const signedDocData = Buffer.from(signatureResult.signed_document, 'base64');
      signedFilename = 'signed_document.pdf';
      await fs.writeFile(signedFilename, signedDocData);
      console.log(`   💾 Saved signed document: ${signedFilename}`);
    }

    // 7. Verify Signature
    if (signedFilename) {
      console.log('\n7️⃣  Verifying signature...');
      const verificationResult = await client.verifySignature(signedFilename);
      
      const isValid = verificationResult.is_valid || false;
      const statusIcon = isValid ? '✅' : '❌';
      console.log(`   ${statusIcon} Signature Valid: ${isValid}`);
      console.log(`   🔍 Verification Level: ${verificationResult.verification_level}`);
      
      const validationDetails = verificationResult.validation_details || {};
      console.log(`   📋 Certificate Valid: ${validationDetails.certificate_valid}`);
      console.log(`   🔒 Signature Intact: ${validationDetails.signature_intact}`);
      console.log(`   ⏱️  Timestamp Valid: ${validationDetails.timestamp_valid}`);
    }

    // 8. Usage Statistics
    console.log('\n8️⃣  Checking usage statistics...');
    try {
      const usageStats = await client.getUsageStats();
      console.log(`   📊 Signatures Created Today: ${usageStats.signatures_created || 0}`);
      console.log(`   ✔️  Signatures Verified Today: ${usageStats.signatures_verified || 0}`);
      console.log(`   📈 API Requests Today: ${usageStats.api_requests || 0}`);
    } catch (error) {
      console.log(`   ⚠️  Usage stats unavailable: ${error.message}`);
    }

    console.log('\n🎉 QES Platform workflow completed successfully!');
    console.log('\n📚 Next Steps:');
    console.log('   • Explore the full API documentation');
    console.log('   • Integrate with your application');
    console.log('   • Test with different signature formats');
    console.log('   • Set up webhook notifications');

  } catch (error) {
    console.log(`\n❌ Error: ${error.message}`);
    
    if (error.stack) {
      console.log('\n🔍 Stack Trace:');
      console.log(error.stack);
    }
    
  } finally {
    // Cleanup
    const filesToCleanup = ['sample_document.pdf', 'signed_document.pdf'];
    
    for (const file of filesToCleanup) {
      try {
        if (await fs.pathExists(file)) {
          await fs.remove(file);
          console.log(`   🧹 Cleaned up: ${file}`);
        }
      } catch (error) {
        console.log(`   ⚠️  Failed to cleanup ${file}: ${error.message}`);
      }
    }
  }
}

/**
 * Check for required dependencies
 */
async function checkDependencies() {
  const requiredPackages = ['axios', 'form-data', 'fs-extra'];
  const missingPackages = [];
  
  for (const pkg of requiredPackages) {
    try {
      require.resolve(pkg);
    } catch (error) {
      missingPackages.push(pkg);
    }
  }
  
  if (missingPackages.length > 0) {
    console.log('❌ Missing required dependencies:');
    missingPackages.forEach(pkg => console.log(`   • ${pkg}`));
    console.log('\nInstall them with:');
    console.log(`   npm install ${missingPackages.join(' ')}`);
    process.exit(1);
  }
}

/**
 * Main function
 */
async function main() {
  console.log('QES Platform JavaScript SDK - Quick Start Example');
  console.log('='.repeat(50));
  console.log();
  console.log('This example demonstrates the complete QES workflow:');
  console.log('• API connectivity check');
  console.log('• Provider listing');
  console.log('• User authentication');
  console.log('• Document signing');
  console.log('• Signature verification');
  console.log();
  
  // Check dependencies
  await checkDependencies();
  
  // Check for environment setup
  const missingVars = [];
  const envVars = ['QES_API_URL', 'QES_API_KEY', 'QES_TENANT_ID'];
  
  for (const varName of envVars) {
    if (!process.env[varName]) {
      missingVars.push(varName);
    }
  }
  
  if (missingVars.length > 0) {
    console.log('⚠️  Missing environment variables for production use:');
    missingVars.forEach(varName => console.log(`   • ${varName}`));
    console.log('\nRunning in demo mode with local development server...');
    console.log('For production, set the above environment variables.');
    console.log();
  }
  
  console.log('Press Enter to start the demonstration...');
  await new Promise(resolve => {
    process.stdin.once('data', resolve);
  });
  
  await demonstrateWorkflow();
}

// Run the demonstration if this file is executed directly
if (require.main === module) {
  main().catch(error => {
    console.error('Fatal error:', error.message);
    process.exit(1);
  });
}

module.exports = {
  QESPlatformClient,
  demonstrateWorkflow,
  createSamplePdf
};